Response validation panic when body uses array circular reference field

I consistently get a panic in a production use case. Here is a minimal reproduction:

```go
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"

	"github.com/pb33f/libopenapi"
	validator "github.com/pb33f/libopenapi-validator"
)

func main() {
	spec := `openapi: 3.1.0
info:
  title: Panic at response validation
  version: 1.0.0
paths:
  /operations:
    delete:
      description: Delete operations
      responses:
        default:
          description: Any response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Error'

components:
  schemas:
    Error:
      type: object
      properties:
        code:
          type: string
        details:
          type: array
          items:
            $ref: '#/components/schemas/Error'
`

	document, err := libopenapi.NewDocument([]byte(spec))
	if err != nil {
		panic(fmt.Sprintf("failed to create new document: %v\n", err))
	}

	_, errs := document.BuildV3Model()
	if len(errs) > 0 {
		for i := range errs {
			fmt.Printf("model error: %e\n", errs[i])
		}
		panic(fmt.Sprintf("failed to create v3 model from document: %d errors reported", len(errs)))
	}

	fmt.Println("Successfully parsed OpenAPI spec")

	oapiValidator, errs := validator.NewValidator(document)
	if errs != nil {
		panic(fmt.Sprintf("failed to create validator: %v", errs))
	}
	if ok, errs := oapiValidator.ValidateDocument(); !ok {
		panic(fmt.Sprintf("document validation errors: %v", errs))
	}

	req := &http.Request{
		Method: http.MethodDelete,
		URL: &url.URL{
			Path: "/operations",
		},
	}
	res := &http.Response{
		StatusCode: http.StatusOK,
		Header: map[string][]string{
			"Content-Type": {"application/json"},
		},
		Body: io.NopCloser(strings.NewReader(`{"code":"abc","details":[{"code":"def"}]}`)),
	}
	if ok, errs := oapiValidator.ValidateHttpResponse(req, res); !ok {
		panic(fmt.Sprintf("response validation erros: %v", errs))
	}
}
```

This results in:

```
Successfully parsed OpenAPI spec
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x8937f1]

goroutine 1 [running]:
github.com/santhosh-tekuri/jsonschema/v6.unevalFrom({0x90be20, 0xc0004dd500}, 0x0, 0x0)
        /home/vanntile/.go/pkg/mod/github.com/santhosh-tekuri/jsonschema/v6@v6.0.1/validator.go:922 +0x131
github.com/santhosh-tekuri/jsonschema/v6.(*Schema).validate(0x0, {0x90be20, 0xc0004dd500}, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/vanntile/.go/pkg/mod/github.com/santhosh-tekuri/jsonschema/v6@v6.0.1/validator.go:25 +0xa9
github.com/santhosh-tekuri/jsonschema/v6.(*Schema).Validate(...)
        /home/vanntile/.go/pkg/mod/github.com/santhosh-tekuri/jsonschema/v6@v6.0.1/validator.go:16
github.com/pb33f/libopenapi-validator/responses.ValidateResponseSchema(0xc0004312c0, 0xc0004d8240, 0xc00038e288, {0xc000394300, 0x97, 0x100}, {0xc0005665a0, 0x82, 0xc000123c20?})
        /home/vanntile/.go/pkg/mod/github.com/pb33f/libopenapi-validator@v0.2.2/responses/validate_response.go:135 +0xebd
github.com/pb33f/libopenapi-validator/responses.(*responseBodyValidator).checkResponseSchema(0xc0004086c0, 0xc0004312c0, 0xc0004d8240, {0xa1cfed?, 0x1?}, 0xc00040b680)
        /home/vanntile/.go/pkg/mod/github.com/pb33f/libopenapi-validator@v0.2.2/responses/validate_body.go:160 +0x2a7
github.com/pb33f/libopenapi-validator/responses.(*responseBodyValidator).ValidateResponseBodyWithPathItem(0xc0004086c0, 0xc0004312c0, 0xc0004d8240, 0x90?, {0xc0002d7b45, 0xb})
        /home/vanntile/.go/pkg/mod/github.com/pb33f/libopenapi-validator@v0.2.2/responses/validate_body.go:92 +0x6ab
github.com/pb33f/libopenapi-validator.(*validator).ValidateHttpResponse(0xc000402910, 0xc0004312c0, 0xc0004d8240)
        /home/vanntile/.go/pkg/mod/github.com/pb33f/libopenapi-validator@v0.2.2/validator.go:124 +0x59
main.main()
        /home/vanntile/src/github.com/vanntile/libopenapi-validator-test/main.go:80 +0x34f
exit status 2
```

Considering that this library does not consider (non-required) circular references errors, and both building a model and validating it pass, I am wondering if this is just a bug or I am missing something.

Tracking down the error, I got to

https://github.com/pb33f/libopenapi-validator/blob/7b4e021e253dae5c215da2c29b5d2862190df7b0/responses/validate_response.go#L131-L132

```
compilation error: json-pointer in "file:///home/vanntile/src/github.com/vanntile/libopenapi-validator-test/response.json#/components/schemas/Error" not found
```

Some error validation would be useful, don't know at the moment how to solve

--- a/go.mod
+++ b/go.mod
@@ -3,7 +3,7 @@ module github.com/pb33f/libopenapi-validator
 go 1.23.0
 
 require (
-	github.com/pb33f/libopenapi v0.18.7
+	github.com/pb33f/libopenapi v0.19.1
 	github.com/santhosh-tekuri/jsonschema/v6 v6.0.1
 	github.com/stretchr/testify v1.10.0
 	github.com/vmware-labs/yaml-jsonpath v0.3.2

--- a/helpers/parameter_utilities.go
+++ b/helpers/parameter_utilities.go
@@ -140,9 +140,12 @@ func ConstructParamMapFromDeepObjectEncoding(values []*QueryParam, sch *base.Sch
 			}
 			// check if schema has additional properties defined as an array
 			if sch != nil && sch.AdditionalProperties != nil &&
-				sch.AdditionalProperties.IsA() &&
-				slices.Contains(sch.AdditionalProperties.A.Schema().Type, Array) {
-				props[v.Property] = rawValues
+				sch.AdditionalProperties.IsA() {
+				s := sch.AdditionalProperties.A.Schema()
+				if s != nil &&
+					slices.Contains(s.Type, Array) {
+					props[v.Property] = rawValues
+				}
 			}
 
 			if len(props) == 0 {

--- a/paths/paths_test.go
+++ b/paths/paths_test.go
@@ -677,7 +677,6 @@ paths:
 }
 
 func TestNewValidator_FindPathWithEncodedArg(t *testing.T) {
-
 	spec := `openapi: 3.1.0
 paths:
   /something/{string_contains_encoded}:

--- a/responses/validate_body.go
+++ b/responses/validate_body.go
@@ -5,6 +5,7 @@ package responses
 
 import (
 	"fmt"
+	"gopkg.in/yaml.v3"
 	"net/http"
 	"strconv"
 	"strings"
@@ -146,20 +147,39 @@ func (v *responseBodyValidator) checkResponseSchema(
 
 				// render the schema inline and perform the intensive work of rendering and converting
 				// this is only performed once per schema and cached in the validator.
-				schema = mediaType.Schema.Schema()
-				renderedInline, _ = schema.RenderInline()
-				renderedJSON, _ = utils.ConvertYAMLtoJSON(renderedInline)
-				v.schemaCache.Store(hash, &schemaCache{
-					schema:         schema,
-					renderedInline: renderedInline,
-					renderedJSON:   renderedJSON,
-				})
+				schemaP := mediaType.Schema
+				marshalled, mErr := schemaP.MarshalYAMLInline()
+
+				if mErr != nil {
+					validationErrors = append(validationErrors, &errors.ValidationError{
+						Reason:            mErr.Error(),
+						Message:           fmt.Sprintf("unable to marshal schema for %s", contentType),
+						ValidationType:    helpers.ResponseBodyValidation,
+						ValidationSubType: helpers.Schema,
+						SpecLine:          mediaType.Schema.GetSchemaKeyNode().Line,
+						SpecCol:           mediaType.Schema.GetSchemaKeyNode().Column,
+						RequestPath:       request.URL.Path,
+						RequestMethod:     request.Method,
+						HowToFix:          "ensure schema is valid and does not contain circular references",
+					})
+				} else {
+					schema = schemaP.Schema()
+					renderedInline, _ = yaml.Marshal(marshalled)
+					renderedJSON, _ = utils.ConvertYAMLtoJSON(renderedInline)
+					v.schemaCache.Store(hash, &schemaCache{
+						schema:         schema,
+						renderedInline: renderedInline,
+						renderedJSON:   renderedJSON,
+					})
+				}
 			}
 
+			if len(renderedInline) > 0 && len(renderedJSON) > 0 && schema != nil {
 			// render the schema, to be used for validation
 			valid, vErrs := ValidateResponseSchema(request, response, schema, renderedInline, renderedJSON, config.WithRegexEngine(v.options.RegexEngine))
 			if !valid {
 				validationErrors = append(validationErrors, vErrs...)
+				}
 			}
 		}
 	}

--- a/responses/validate_body.go
+++ b/responses/validate_body.go
@@ -5,7 +5,6 @@ package responses
 
 import (
 	"fmt"
-	"gopkg.in/yaml.v3"
 	"net/http"
 	"strconv"
 	"strings"
@@ -13,6 +12,7 @@ import (
 	"github.com/pb33f/libopenapi/datamodel/high/base"
 	"github.com/pb33f/libopenapi/orderedmap"
 	"github.com/pb33f/libopenapi/utils"
+	"gopkg.in/yaml.v3"
 
 	v3 "github.com/pb33f/libopenapi/datamodel/high/v3"
 
@@ -175,10 +175,10 @@ func (v *responseBodyValidator) checkResponseSchema(
 			}
 
 			if len(renderedInline) > 0 && len(renderedJSON) > 0 && schema != nil {
-			// render the schema, to be used for validation
-			valid, vErrs := ValidateResponseSchema(request, response, schema, renderedInline, renderedJSON, config.WithRegexEngine(v.options.RegexEngine))
-			if !valid {
-				validationErrors = append(validationErrors, vErrs...)
+				// render the schema, to be used for validation
+				valid, vErrs := ValidateResponseSchema(request, response, schema, renderedInline, renderedJSON, config.WithRegexEngine(v.options.RegexEngine))
+				if !valid {
+					validationErrors = append(validationErrors, vErrs...)
 				}
 			}
 		}

--- a/responses/validate_body_test.go
+++ b/responses/validate_body_test.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"net/http"
 	"net/http/httptest"
+	"net/url"
 	"testing"
 
 	"github.com/pb33f/libopenapi"
@@ -1299,6 +1300,69 @@ paths:
 
 	assert.True(t, valid)
 	assert.Len(t, errors, 0)
+}
+
+// https://github.com/pb33f/libopenapi-validator/issues/107
+// https://github.com/pb33f/libopenapi-validator/issues/103
+func TestNewValidator_TestCircularRefsInValidation_Response(t *testing.T) {
+	spec := `openapi: 3.1.0
+info:
+  title: Panic at response validation
+  version: 1.0.0
+paths:
+  /operations:
+    delete:
+      description: Delete operations
+      responses:
+        default:
+          description: Any response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+
+components:
+  schemas:
+    Error:
+      type: object
+      properties:
+        code:
+          type: string
+        details:
+          type: array
+          items:
+            $ref: '#/components/schemas/Error'`
+
+	doc, _ := libopenapi.NewDocument([]byte(spec))
+
+	m, _ := doc.BuildV3Model()
+	v := NewResponseBodyValidator(&m.Model)
+
+	req := &http.Request{
+		Method: http.MethodDelete,
+		URL: &url.URL{
+			Path: "/operations",
+		},
+	}
+	// simulate a request/response
+	res := httptest.NewRecorder()
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set(helpers.ContentTypeHeader, helpers.JSONContentType)
+		w.WriteHeader(http.StatusOK)
+		_, _ = w.Write(nil)
+	}
+
+	// fire the request
+	handler(res, req)
+
+	// record response
+	response := res.Result()
+
+	valid, errors := v.ValidateResponseBody(req, response)
+
+	assert.False(t, valid)
+	assert.Len(t, errors, 1)
+	assert.Equal(t, "cannot render circular reference: #/components/schemas/Error", errors[0].Reason)
 }
 
 type errorReader struct{}

--- a/validator_test.go
+++ b/validator_test.go
@@ -1545,6 +1545,7 @@ paths:
 }
 
 // https://github.com/pb33f/libopenapi-validator/issues/107
+// https://github.com/pb33f/libopenapi-validator/issues/103
 func TestNewValidator_TestCircularRefsInValidation_Request(t *testing.T) {
 
 	spec := `openapi: 3.1.0
@@ -1622,6 +1623,7 @@ components:
 }
 
 // https://github.com/pb33f/libopenapi-validator/issues/107
+// https://github.com/pb33f/libopenapi-validator/issues/103
 func TestNewValidator_TestCircularRefsInValidation_Response(t *testing.T) {
 
 	spec := `openapi: 3.1.0

--- a/validator_test.go
+++ b/validator_test.go
@@ -6,9 +6,13 @@ package validator
 import (
 	"bytes"
 	"encoding/json"
+	"fmt"
+	"io"
 	"net/http"
 	"net/http/httptest"
+	"net/url"
 	"os"
+	"strings"
 	"sync"
 	"testing"
 
@@ -1538,4 +1542,151 @@ paths:
 
 	assert.True(t, valid)
 	assert.Len(t, errors, 0)
+}
+
+// https://github.com/pb33f/libopenapi-validator/issues/107
+func TestNewValidator_TestCircularRefsInValidation_Request(t *testing.T) {
+
+	spec := `openapi: 3.1.0
+info:
+  title: Panic at response validation
+  version: 1.0.0
+paths:
+  /operations:
+    delete:
+      description: Delete operations
+      responses:
+        default:
+          description: Any response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+
+components:
+  schemas:
+    Error:
+      type: object
+      properties:
+        code:
+          type: string
+        details:
+          type: array
+          items:
+            $ref: '#/components/schemas/Error'`
+
+	document, err := libopenapi.NewDocument([]byte(spec))
+	if err != nil {
+		panic(fmt.Sprintf("failed to create new document: %v\n", err))
+	}
+
+	model, errs := document.BuildV3Model()
+	circ := model.Index.GetCircularReferences()
+	fmt.Println("Circular references: ", len(circ))
+
+	if len(errs) > 0 {
+		for i := range errs {
+			fmt.Printf("model error: %e\n", errs[i])
+		}
+		panic(fmt.Sprintf("failed to create v3 model from document: %d errors reported", len(errs)))
+	}
+
+	fmt.Println("Successfully parsed OpenAPI spec")
+
+	oapiValidator, errs := NewValidator(document)
+	if errs != nil {
+		panic(fmt.Sprintf("failed to create validator: %v", errs))
+	}
+	if ok, errs := oapiValidator.ValidateDocument(); !ok {
+		panic(fmt.Sprintf("document validation errors: %v", errs))
+	}
+
+	req := &http.Request{
+		Method: http.MethodDelete,
+		URL: &url.URL{
+			Path: "/operations",
+		},
+	}
+	res := &http.Response{
+		StatusCode: http.StatusOK,
+		Header: map[string][]string{
+			"Content-Type": {"application/json"},
+		},
+		Body: io.NopCloser(strings.NewReader(`{"code":"abc","details":[{"code":"def"}]}`)),
+	}
+	if ok, errs := oapiValidator.ValidateHttpResponse(req, res); !ok {
+		assert.Equal(t, 1, len(errs))
+		assert.Equal(t, "cannot render circular reference: #/components/schemas/Error", errs[0].Reason)
+
+	}
+}
+
+// https://github.com/pb33f/libopenapi-validator/issues/107
+func TestNewValidator_TestCircularRefsInValidation_Response(t *testing.T) {
+
+	spec := `openapi: 3.1.0
+info:
+  title: Panic at response validation
+  version: 1.0.0
+paths:
+  /operations:
+    delete:
+      description: Delete operations
+      responses:
+        default:
+          description: Any response
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Error'
+
+components:
+  schemas:
+    Error:
+      type: object
+      properties:
+        code:
+          type: string
+        details:
+          type: array
+          items:
+            $ref: '#/components/schemas/Error'`
+
+	document, err := libopenapi.NewDocument([]byte(spec))
+	if err != nil {
+		panic(fmt.Sprintf("failed to create new document: %v\n", err))
+	}
+
+	model, errs := document.BuildV3Model()
+	circ := model.Index.GetCircularReferences()
+	fmt.Println("Circular references: ", len(circ))
+
+	if len(errs) > 0 {
+		for i := range errs {
+			fmt.Printf("model error: %e\n", errs[i])
+		}
+		panic(fmt.Sprintf("failed to create v3 model from document: %d errors reported", len(errs)))
+	}
+
+	fmt.Println("Successfully parsed OpenAPI spec")
+
+	oapiValidator, errs := NewValidator(document)
+	if errs != nil {
+		panic(fmt.Sprintf("failed to create validator: %v", errs))
+	}
+	if ok, errs := oapiValidator.ValidateDocument(); !ok {
+		panic(fmt.Sprintf("document validation errors: %v", errs))
+	}
+
+	req := &http.Request{
+		Method: http.MethodDelete,
+		URL: &url.URL{
+			Path: "/operations",
+		},
+	}
+	if ok, errs := oapiValidator.ValidateHttpRequest(req); !ok {
+		assert.Equal(t, 1, len(errs))
+		assert.Equal(t, "cannot render circular reference: #/components/schemas/Error", errs[0].Reason)
+
+	}
 }

--- a/validator_test.go
+++ b/validator_test.go
@@ -1547,7 +1547,6 @@ paths:
 // https://github.com/pb33f/libopenapi-validator/issues/107
 // https://github.com/pb33f/libopenapi-validator/issues/103
 func TestNewValidator_TestCircularRefsInValidation_Request(t *testing.T) {
-
 	spec := `openapi: 3.1.0
 info:
   title: Panic at response validation
@@ -1625,7 +1624,6 @@ components:
 // https://github.com/pb33f/libopenapi-validator/issues/107
 // https://github.com/pb33f/libopenapi-validator/issues/103
 func TestNewValidator_TestCircularRefsInValidation_Response(t *testing.T) {
-
 	spec := `openapi: 3.1.0
 info:
   title: Panic at response validation

